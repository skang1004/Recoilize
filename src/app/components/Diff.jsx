// renders the difference between the most recent state change and the previous
import React, { useState } from 'react';
import { diff, formatters } from 'jsondiffpatch';
import ReactHtmlParser from 'react-html-parser';

const Diff = ({ oldSnap, newSnap }) => {
  //
  const [rawToggle, setRawToggle] = useState(false);
  // diffing between oldSnap && newSnap
  const delta = diff(oldSnap, newSnap);
  // string of html with comparisons
  const html = formatters.html.format(delta, oldSnap);
  // conditionally render changes or not
  rawToggle ? formatters.html.showUnchanged() : formatters.html.hideUnchanged();

  return (
    <div className='Diff'>
      <button
        onClick={() => {
          setRawToggle(!rawToggle);
        }}
      >
        Toggle Raw
      </button>
      {ReactHtmlParser(html)}
    </div>
  );
};

export default Diff;
