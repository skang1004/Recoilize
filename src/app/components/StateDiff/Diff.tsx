import React, {useState, useEffect} from 'react';
import {diff, formatters} from 'jsondiffpatch';
import ReactHtmlParser from 'react-html-parser';
import {filteredSnapshot} from '../../../types';

interface DiffProps {
  // snapshot at index [curRender -1]
  filteredPrevSnap: filteredSnapshot;
  // snapshot at index [curRender]
  filteredCurSnap: filteredSnapshot;
  //length of snapshotHistory
}

// renders the difference between the most recent state change and the previous
const Diff: React.FC<DiffProps> = ({filteredPrevSnap, filteredCurSnap}) => {
  // console.log('this is the filteredSnap ', filteredCurSnap);
  // useState hook to update the toggle of showUnchanged or hideUnchanged
  const [rawToggle, setRawToggle] = useState(false);
  // diffing between filteredPrevSnap && filteredCurSnap
  const delta = diff(filteredPrevSnap, filteredCurSnap);
  // console.log('this is the delta ', delta);
  // string of html with comparisons
  const html = formatters.html.format(delta, filteredPrevSnap);
  // conditionally render changes or not based on rawToggle bool
  formatters.html.showUnchanged(rawToggle);

  return (
    <div className="Diff">
      <div className="toggleDiv">
        <button
          className="rawToggle"
          style={{color: rawToggle ? '#E6E6E6' : '#989898'}}
          onClick={() => {
            setRawToggle(!rawToggle);
          }}>
          Raw
        </button>
      </div>
      {ReactHtmlParser(html)}
    </div>
  );
};

export default Diff;
