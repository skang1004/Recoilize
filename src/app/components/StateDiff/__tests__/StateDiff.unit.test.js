import React from 'react';
import ReactDOM from 'react-dom';
import Diff from '../Diff';

import {
  render,
  cleanup,
  getByText,
  fireEvent,
  debug,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import {
  filteredPrevSnapMock,
  filteredCurSnapMock,
} from '../../../../../mock/snapshot.js';

afterEach(cleanup);

const mockProps = {
  filteredPrevSnap: filteredPrevSnapMock,
  filteredCurSnap: filteredCurSnapMock,
};
describe('Diff Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Diff />, div);
  });
});

describe('Raw Toggle Button', () => {
  beforeAll(cleanup);
  afterAll(cleanup);

  it('raw button should be color #989898 at initial render', () => {
    const {getByText} = render(<Diff />);
    expect(getByText('Raw')).toHaveStyle('color: #989898');
  });

  it('raw button should change to color #E6E6E6 after clicked', () => {
    const {getByText} = render(<Diff />);
    fireEvent.click(getByText('Raw'));
    expect(getByText('Raw')).toHaveStyle('color: #E6E6E6');
  });
});

it('should match snapshot when no props are passed in', () => {
  const {asFragment} = render(<Diff />);
  expect(asFragment()).toMatchSnapshot();
});

it('should match snapshot when props are passed in', () => {
  const {asFragment} = render(<Diff {...mockProps} />);
  expect(asFragment()).toMatchSnapshot();
});
