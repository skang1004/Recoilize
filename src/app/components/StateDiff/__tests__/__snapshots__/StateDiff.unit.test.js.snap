// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot when no props are passed in 1`] = `
<DocumentFragment>
  <div
    class="Diff"
  >
    <div
      class="toggleDiv"
    >
      <button
        class="rawToggle"
        style="color: rgb(152, 152, 152);"
      >
        Raw
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot when props are passed in 1`] = `
<DocumentFragment>
  <div
    class="Diff"
  >
    <div
      class="toggleDiv"
    >
      <button
        class="rawToggle"
        style="color: rgb(152, 152, 152);"
      >
        Raw
      </button>
    </div>
    <div
      class="jsondiffpatch-delta jsondiffpatch-node jsondiffpatch-child-node-type-object"
    >
      <ul
        class="jsondiffpatch-node jsondiffpatch-node-type-object"
      >
        <li
          class="jsondiffpatch-unchanged"
          data-key="dummyAtom1"
        >
          <div
            class="jsondiffpatch-property-name"
          >
            dummyAtom1
          </div>
          <div
            class="jsondiffpatch-value"
          >
            <pre>
              {
  "contents": {
    "hello": [],
    "hi": []
  },
  "nodeDeps": [],
  "nodeToNodeSubscriptions": [],
  "type": "RecoilState"
}
            </pre>
          </div>
        </li>
        <li
          class="jsondiffpatch-unchanged"
          data-key="listState"
        >
          <div
            class="jsondiffpatch-property-name"
          >
            listState
          </div>
          <div
            class="jsondiffpatch-value"
          >
            <pre>
              {
  "contents": [
    {
      "text": "list item"
    },
    {
      "text": "list item"
    },
    {
      "text": "list item"
    }
  ],
  "nodeDeps": [],
  "nodeToNodeSubscriptions": [
    "selectorTest",
    "stateLengths"
  ],
  "type": "RecoilState"
}
            </pre>
          </div>
        </li>
        <li
          class="jsondiffpatch-node jsondiffpatch-child-node-type-object"
          data-key="listState2"
        >
          <div
            class="jsondiffpatch-property-name"
          >
            listState2
          </div>
          <ul
            class="jsondiffpatch-node jsondiffpatch-node-type-object"
          >
            <li
              class="jsondiffpatch-node jsondiffpatch-child-node-type-array"
              data-key="contents"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                contents
              </div>
              <ul
                class="jsondiffpatch-node jsondiffpatch-node-type-array"
              >
                <li
                  class="jsondiffpatch-unchanged"
                  data-key="0"
                >
                  <div
                    class="jsondiffpatch-property-name"
                  >
                    0
                  </div>
                  <div
                    class="jsondiffpatch-value"
                  >
                    <pre>
                      {
  "text": "list item"
}
                    </pre>
                  </div>
                </li>
                <li
                  class="jsondiffpatch-unchanged"
                  data-key="1"
                >
                  <div
                    class="jsondiffpatch-property-name"
                  >
                    1
                  </div>
                  <div
                    class="jsondiffpatch-value"
                  >
                    <pre>
                      {
  "text": "list item"
}
                    </pre>
                  </div>
                </li>
                <li
                  class="jsondiffpatch-unchanged"
                  data-key="2"
                >
                  <div
                    class="jsondiffpatch-property-name"
                  >
                    2
                  </div>
                  <div
                    class="jsondiffpatch-value"
                  >
                    <pre>
                      {
  "text": "list item"
}
                    </pre>
                  </div>
                </li>
                <li
                  class="jsondiffpatch-deleted"
                  data-key="3"
                >
                  <div
                    class="jsondiffpatch-property-name"
                  >
                    3
                  </div>
                  <div
                    class="jsondiffpatch-value"
                  >
                    <pre>
                      {
  "text": "list item"
}
                    </pre>
                  </div>
                </li>
              </ul>
            </li>
            <li
              class="jsondiffpatch-unchanged"
              data-key="nodeDeps"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                nodeDeps
              </div>
              <div
                class="jsondiffpatch-value"
              >
                <pre>
                  []
                </pre>
              </div>
            </li>
            <li
              class="jsondiffpatch-unchanged"
              data-key="nodeToNodeSubscriptions"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                nodeToNodeSubscriptions
              </div>
              <div
                class="jsondiffpatch-value"
              >
                <pre>
                  [
  "stateLengths"
]
                </pre>
              </div>
            </li>
            <li
              class="jsondiffpatch-unchanged"
              data-key="type"
            >
              <div
                class="jsondiffpatch-property-name"
              >
                type
              </div>
              <div
                class="jsondiffpatch-value"
              >
                <pre>
                  "RecoilState"
                </pre>
              </div>
            </li>
          </ul>
        </li>
        <li
          class="jsondiffpatch-unchanged"
          data-key="selectorTest"
        >
          <div
            class="jsondiffpatch-property-name"
          >
            selectorTest
          </div>
          <div
            class="jsondiffpatch-value"
          >
            <pre>
              {
  "contents": "test",
  "nodeDeps": [
    "listState"
  ],
  "nodeToNodeSubscriptions": [],
  "type": "RecoilValueReadOnly"
}
            </pre>
          </div>
        </li>
        <li
          class="jsondiffpatch-unchanged"
          data-key="stateLengths"
        >
          <div
            class="jsondiffpatch-property-name"
          >
            stateLengths
          </div>
          <div
            class="jsondiffpatch-value"
          >
            <pre>
              {
  "contents": 6,
  "nodeDeps": [
    "listState",
    "listState2"
  ],
  "nodeToNodeSubscriptions": [],
  "type": "RecoilValueReadOnly"
}
            </pre>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
